stl/inc/type_traits

```diff
@@ -1699,6 +1710,19 @@ template <class _Callable, class _Ty1, class... _Types2>
 using _Decltype_invoke_nonzero = decltype(
     _Invoker1<_Callable, _Ty1>::_Call(_STD declval<_Callable>(), _STD declval<_Ty1>(), _STD declval<_Types2>()...));
 
+template <class _Void, class _Rx, class _Callable, class _Ty1, class... _Types2>
+struct _Invocable_r_nonzero_nonvoid : false_type {};
+
+template <class _Rx, class _Callable, class _Ty1, class... _Types2>
+struct _Invocable_r_nonzero_nonvoid<void_t<decltype(_Implicitly_convert_to<_Rx>(_Invoker1<_Callable, _Ty1>::_Call(
+                                        _STD declval<_Callable>(), _STD declval<_Ty1>(), _STD declval<_Types2>()...)))>,

```

Comment:
No change requested - I thought that repeating the call expression (outside of `noexcept` where that's unavoidable) was unnecessarily verbose and harmful to throughput, so I tried to centralize that. My attempt passed for C1XX but failed for Clang and EDG, probably because I don't really understand C++17 deferred temporary materialization. You can see my attempt at https://github.com/StephanTLavavej/STL/tree/invocable_r-FAILED-SIMPLIFICATION although it's probably not salvageable.

Reply:
Maybe it's because `std::declval` always returns `X&` or `X&&`, never a prvalue `X`?

Reply:
> Maybe it's because `std::declval` always returns `X&` or `X&&`, never a prvalue `X`?

Yes. The crux of this entire issue is that `declval<T>()` is always a glvalue and never a prvalue. If I add another utility function template whose return type is exactly its template parameter, then I think we can use that to avoid reevaluating the call expression for `_Is_invocable_r` and `_Is_nothrow_invocable_r`.

Reply:
Done, simplified slightly, looked at your changes and took the rest upon realizing they were roughly similar.